xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,345 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.SHUTDOWN;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import io.grpc.Attributes;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.InternalLogId;
+import io.grpc.LoadBalancer;
+import io.grpc.Status;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A {@link LoadBalancer} that provides consistent hashing based load balancing to upstream hosts.
+ * It implements the "Ketama" hashing that maps hosts onto a circle (the "ring") by hashing its
+ * addresses. Each request is routed to a host by hashing some property of the request and finding
+ * the neaerest corresponding host clockwise around the ring. Each host is placed on the ring some
+ * number of times proportional to its weight. With the ring partitioned appropriately, the
+ * addition or removal of one host from a set of N hosts will affect only 1/N requests.
+ */
+final class RingHashLoadBalancer extends LoadBalancer {
+  private static final Attributes.Key<AtomicReference<ConnectivityStateInfo>> STATE_INFO =
+      Attributes.Key.create("state-info");
+
+  private final XdsLogger logger = XdsLogger.withLogId(
+      InternalLogId.allocate("ring_hash_lb", null));
+  private final XxHash64 hashFunc = XxHash64.INSTANCE;
+  private final ConcurrentMap<EquivalentAddressGroup, Subchannel> subchannels =
+      Maps.newConcurrentMap();
+  private final Helper helper;
+
+  private ConnectivityState currentState;
+  private SubchannelPicker currentPicker;
+  private boolean subchannelEnteredReady;
+
+  RingHashLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    List<EquivalentAddressGroup> addrList = resolvedAddresses.getAddresses();
+    if (addrList.isEmpty()) {
+      handleNameResolutionError(Status.UNAVAILABLE.withDescription("No server addresses found"));
+      return;
+    }
+    Map<EquivalentAddressGroup, EquivalentAddressGroup> latestAddrs = stripAttrs(addrList);
+    Set<EquivalentAddressGroup> removedAddrs =
+        Sets.newHashSet(Sets.difference(subchannels.keySet(), latestAddrs.keySet()));
```

Comment:
[`Sets.difference(...)`](https://guava.dev/releases/22.0/api/docs/com/google/common/collect/Sets.html#difference-java.util.Set-java.util.Set-) gives you an _unmodifiable **view**_ only. That is, if either `subchannels` or `latestAddrs` is mutated, it will be seen by the view. We are iterating through the removed addresses to remove subchannelss.

Reply:
Ah, I see `subchannels` is mutated later. So it makes sense. But I didn't see `ConcurrentModificationException` running the tests without `Sets.newHashSet()`.

Reply:
Maybe because `subchannels` is a ConcurrentMap? I don't know exactly how the SetView works, but based on its doc

> The returned set contains all elements that are contained by set1 and not contained by set2. set2 may also contain elements not present in set1; these are simply ignored. The iteration order of the returned set matches that of set1.

It's probably based on set1, which is `subchannels` in this case.